### ISSUE TITLE
[execution_dependencies] Incorrect order of execution

### ISSUE THREAD

**Issue by aditya95sriram** [ipython-contrib/jupyter_contrib_nbextensions#1423]
[Here](https://gist.github.com/aditya95sriram/487b447cd72bc2214e745b43fb8e8eb0) is a simplified example of a notebook I was working on, in which the execution_dependencies extension was failing to run the dependencies in the correct order.

The code cells are structured as shown below (lower cells dependent on higher cells)
```
   ╔═══╗ ╔═══╗  
   ║ a ║ ║ b ║
   ╚═╤═╝ ╚═╤═╝
     └─┐ ┌─┘
      ╔╧═╧╗
      ║ r ║
      ╚╤╤╤╝
  ┌────┘│└────┐
╔═╧═╗ ╔═╧═╗ ╔═╧═╗
║ h ║ ║ e ║ ║ i ║
╚═╤═╝ ╚═╤═╝ ╚═╤═╝
  └───┐ │ ┌───┘
     ╔╧═╧═╧╗
     ║start║
     ╚═════╝
```
When I execute the `start` cell, I see that first cell to be executed is the `h` cell, which then throws an error because it depends on cell `r` which hasn't been run yet. When I peeked into the console, I saw that there was a log message saying "executing dependency cells in order `h,r,i,e`". 
As can be seen, the correct execution order should be: `a,b,r,h,e,i,start` (or some permutation of the cells on the same layer).

Digging into the code a bit deeper (thanks @benelot for already including commented `console.logs`), I found out that this problem is arising because the in-degrees aren't being computed properly. See the table below for a comparison of `in_degree` 

Cell | Computed in-degree | Correct in-degree
----|----|---
a | 2 | **1**
b | 2 | **1**
r | 2 | **3**
h | 1 | 1 
e | 1 | 1 
i | 1 | 1
start | 0 | 0

**Comment by benelot** [comment]
Thanks for that perfect analysis. Do you by chance have the fix for the calculation of the in_degree? I will have to dig into the code again and I think you know more about it than I do right now.

**Comment by benelot** [comment]
On a side note, how did you draw that very neat graph?

**Comment by aditya95sriram** [comment]
> Thanks for that perfect analysis. Do you by chance have the fix for the calculation of the in_degree? I will have to dig into the code again and I think you know more about it than I do right now.

Glad you asked, I did some more digging(breakpoints and chrome debugger) and I feel that the method you use to terminate the in-degree calculation is slightly flawed. I saw that the same cell (`r` in this case) was considered more than once during the in-degree calculation loop (because `h,e,i` all depend on `r`) and it contributed twice to `processed_nodes`, which is why the loop terminated prematurely. I see an easy non-intrusive fix for this particular issue, which is to introduce an object which keeps track of "visited" nodes and whenever an already visited node is encountered during the loop, simply skip over it.

Let me know if that sounds confusing, I can probably also create a pull-request with what I have in mind, if it helps. 


> Edit: On second thought, I think the indegree is not wrong. Since you do want to run the start cell last, the indegree is calculated top down, not bottom up in your graph.

Sorry, I don't understand this. Just to be clear, the in-degree of a node refers to the number of other nodes that depend on it and not the other way around, right ?

**Comment by aditya95sriram** [comment]
> On a side note, how did you draw that very neat graph?

The basic idea is to insert ascii characters via alt codes [(see this page)](https://www.alt-codes.net/). In this case I created the graph manually, by inserting all the required characters into my text file (kind of like a snippet bank or toolbar) and then copy-pasting from there [(example)](https://pastebin.com/3KqSE9P8). Also, Notepad++'s column editing mode comes in handy at times. 

Currently, it's a little too much effort, but I'm pretty sure there should be an online tool to create such simple lo-fi graphs. Either ways, I'm planning on creating one such tool myself, I'll let you know once it's done 😃

**Comment by benelot** [comment]
> Sorry, I don't understand this. Just to be clear, the in-degree of a node refers to the number of other nodes that depend on it and not the other way around, right ?

Yes, but shouldn't a and b be independent? As far as I can tell r depends on a and b and not the other way around. The calculation is flawed indeed if above numbers come out.

Should the in degrees not be the following?
Cell | Computed in-degree | Correct in-degree
-- | -- | --
a | 2 | 0
b | 2 | 0
r | 2 | 2
h | 1 | 1
e | 1 | 1
i | 1 | 1
start | 0 | 3

My code below the indegree calculation then just picks one of the 0 indegree nodes, decreases the indegree of the nodes that depend on it and then executes it. That would make sense with the above indegrees and result in the desired order, but would not work with your indegrees, right?

**Comment by aditya95sriram** [comment]
That table does make much more sense in the context of the code that follows it. And I get it now, let `x` be dependent on `y`, then you orient the edge as `x ← y` and then the table you've shown is the correct in-degree table. And then your logic works properly because you pick the edges with in-degree 0 (a.k.a no cells depend on this cell) and insert it into the processing order. 

Albeit, the visited_nodes fix might still be required. Do let me know if you need help.

**Comment by benelot** [comment]
Haha, right. In the context of dependencies, there is the notion to have the arrow from the node depending on to the node it depends on. But That might actually be confusing. If you think my readme does not state that properly, please tell me and we improve it.

I currently have problems installing my own extension since it says it is incompatible with the newest jupyter notebook.

**Comment by aditya95sriram** [comment]
> Haha, right. In the context of dependencies, there is the notion to have the arrow from the node depending on to the node it depends on. But That might actually be confusing. If you think my readme does not state that properly, please tell me and we improve it.

This orientation stuff is only of concern for developers so I don't think it needs to be in the readme. Maybe a comment in the code itself could resolve this confusion. Also, another thing adding to this confusion is the fact that the dependencies are denoted by the tag (`=>y`) which as you said makes more semantic sense but coupled with the term "in-degree" it's slightly confusing to developers.

> I currently have problems installing my own extension since it says it is incompatible with the newest jupyter notebook.

I can try my hand at getting the code to generate in-degrees as per your last table. Is there anything else you had in mind ?

**Comment by benelot** [comment]
Maybe I could just make it work for the moment. Let me see. I will give you an update.

**Comment by benelot** [comment]
I think the in-degree calculation can be massively simlified. Since the notation is the way it is, I write #start, =>A, =>B, =>C. So counting the number of in-degrees is as simple as counting the number of defined dependencies. I originally had the inverted notation (the one that you expected), that is why I calculate it so strangely. So could you test for me if the following makes it work:

Replace this:
```
                           for(var i=0, dep_qty=dep_graph[id].length; i < dep_qty; i++) {
                                var dep = dep_graph[id][i];
                          //    console.log('      dep: ', dep);
                                in_degree[id] = in_degree[id] || 0;
                                in_degree[dep] = in_degree[dep] === undefined ? 1 : ++in_degree[dep];
                                processing_queue.unshift(dep);
                            }
```

with 

```
   in_degree[id] = dep_graph[id].length
   //    console.log('      deps: ', dep_graph[id]);
   processing_queue.unshift(dep_graph[id])
```

Like that we still have the ability to find cycles, but also we store the right in degrees if I am not mistaken. Please test this for me, I am just compiling this in my head. :D Just check if the simple cases still work like no dependencies, a single dependency and a circular dependency.

If it works, you can pull-request it.

**Comment by aditya95sriram** [comment]
Okay, that does make sense. Don't know how I casually accepted the quirky code that was there earlier. This one's so much more straightforward. I tried it out on my example with two changes 
1. replaced `extend`  with `concat` (`a = a.concat(b)` to be precise)
2. had to add in visited code as you suspected, because the same node was being processed multiple times (this seems simpler and less intrusive than converting `processing_queue` into a set)

and the `in_degree` calculation works exactly as you want it to, but the `processing_order` is still computed incorrectly as `h,r,i,e`. I'm not sure why, but I'm looking into it.

**Comment by aditya95sriram** [comment]
Also, I have a feeling that we don't need to maintain a separate `visited` object and we can simply check if the `in_degree[id]` is defined or not to detect if it has been processed. But I'm postponing optimizations for later, my first priority being - getting the execution order correctly.

**Comment by benelot** [comment]
Sorry for deleting the comment above, here it is again: 
I think the in-degree calculation can be massively simplified. Since the notation is the way it is, I write #start, =>A, =>B, =>C. So counting the number of in-degrees is as simple as counting the number of defined dependencies. I originally had the inverted notation (the one that you expected), that is why I calculate it so strangely. So could you test for me if the following makes it work:

Replace this:
```
                           for(var i=0, dep_qty=dep_graph[id].length; i < dep_qty; i++) {
                                var dep = dep_graph[id][i];
                          //    console.log('      dep: ', dep);
                                in_degree[id] = in_degree[id] || 0;
                                in_degree[dep] = in_degree[dep] === undefined ? 1 : ++in_degree[dep];
                                processing_queue.unshift(dep);
                            }
```

with 

```
   in_degree[id] = dep_graph[id].length
   //    console.log('      deps: ', dep_graph[id]);
   processing_queue.extend(dep_graph[id])
```

Like that we still have the ability to find cycles, but also we store the right in degrees if I am not mistaken. Please test this for me, I am just compiling this in my head. :D Just check if the simple cases still work like no dependencies, a single dependency and a circular dependency.

If it works, you can pull-request it. 

Edit: I replaced unshift with extend since we add one list to the other. Still not so sure if it is the right method, javascript is crazy...

Edit2: Oh and of course every node should only be processed once, otherwise it will always consider the graph to be cyclic. So you will have turn the processing_queue into a processing_set, so that it is not possible to add some duplicates to it. Or you will have to add the visited_list so that you can skip it and not increment the processed_nodes variable.

**Comment by benelot** [comment]
I found the solution I believe. We can simplify the whole first while loop I think:

```
                        var processed_nodes = 0;
                        var in_degree = {};                                                                    // ...collect in-degrees of nodes
                        while(processing_queue.length > 0 && processed_nodes < Object.keys(dep_graph).length) {// ...stay processing deps while the queue contains nodes and the processed nodes are below total node quantity
                            var id = processing_queue.shift();                                                 // .....pop front of queue and front-push it to the processing order
                          //console.log("ID: ", id);
                            for(var i=0, dep_qty=dep_graph[id].length; i < dep_qty; i++) {
                                var dep = dep_graph[id][i];
                          //    console.log('      dep: ', dep);
                                in_degree[id] = in_degree[id] || 0;
                                in_degree[dep] = in_degree[dep] === undefined ? 1 : ++in_degree[dep];
                                processing_queue.unshift(dep);
                            }
                            processed_nodes++;
                        }
```

can be replaced by

```
                        var processed_nodes = 0;
                        var in_degree = {};                                                                    // ...collect in-degrees of nodes
                        for(var i=0; i < cell_ids.length; i++) {                                               // ... process all nodes
                            var id = cell_ids[i];
                           //console.log("ID: ", id);
                           in_degree[id] = dep_graph[id].length
						   //    console.log('      deps: ', dep_graph[id]);
                        }
```

**Comment by benelot** [comment]
This avoids to get into a cycle because it just goes through every node exactly once and assigns the correct in-degree number.

The second loop then finds cycles and aborts if necessary, otherwise it just adds a node into the processing exactly once because it is only done if the in-degree counter is down to 0.

**Comment by aditya95sriram** [comment]
I tried your latest suggestion and the variable `cell_ids` is not in scope, so that will have to be redefined. Rather than that, why can we not just loop over the keys of `dep_graph` ?

Also, I think there another minor issue with this approach. The `dep_graph` stores the dependencies of a particular node, but when doing the topological sort, we want the other direction, i.e. we want the list of nodes that depend on a particular node, so that we can go and decrement their in-degree. So, either we loop over all nodes within each outer loop to find the nodes that depend on a given node or we rethink how we are storing `dep_graph`.

**Comment by aditya95sriram** [comment]
Okay, how about this
Replace
```js
while(processing_queue.length > 0 && processed_nodes < Object.keys(dep_graph).length) {// ...stay processing deps while the queue contains nodes and the processed nodes are below total node quantity
  //console.log("ID: ", id);
    //console.log(processing_order, JSON.parse(JSON.stringify(in_degree)));
    for(var i=0, dep_qty=dep_graph[id].length; i < dep_qty; i++) {                     // ......iterate over dependent nodes of current id and decrease their in-degree by 1
        var dep = dep_graph[id][i];
      //console.log('      dep: ', dep);
        in_degree[dep]--;
        if(in_degree[dep] == 0) {                                                      // ......queue dependency if in-degree is 0
            processing_queue.unshift(dep);
        }
    }
    processed_nodes++;
}
```
with this

```js
var all_nodes = Object.keys(dep_graph);
while(processing_queue.length > 0 && processed_nodes < Object.keys(dep_graph).length) {// ...stay processing deps while the queue contains nodes and the processed nodes are below total node quantity
    var id = processing_queue.shift();                                                 // .....pop front of queue and front-push it to the processing order
    if (in_degree[id] == 0) {                                                          // ...no dependencies remaining, add to end of order
        processing_order.push(id);
        processed_nodes++;
        for (var i=0; i < all_nodes.length; i++) {                                     // ... loop over all nodes
            if (dep_graph[all_nodes[i]].indexOf(id) > -1) {                            // ... decrement indegree of nodes depending on id
                in_degree[all_nodes[i]]--;
            }
        }
    } else {
        processing_queue.unshift(id);                                                   // node not ready to be added to processing_order, put it back in the queue
        processing_queue = dep_graph[id].concat(processing_queue);                      // add all its dependencies to the queue too
    }
}
```

I tested this with the `in_degree` code from [here](https://github.com/ipython-contrib/jupyter_contrib_nbextensions/issues/1423#issuecomment-501204965) and it seems to be working fine. 

Let me know what you think, in the meanwhile, I'll try it out on other examples.

**Comment by benelot** [comment]
You are right, we definitely need the inverse dependency graph.

How about inverting the dep graph here?
```
                        var inv_dep_graph = {}
                        var in_degree = {};                                                                    // ...collect in-degrees of nodes
                        for(var i=0; i < cell_ids.length; i++) {                                               // ... process all nodes
                            var id = cell_ids[i];
                           //console.log("ID: ", id);
                           in_degree[id] = dep_graph[id].length
			   //    console.log('      deps: ', dep_graph[id]);
                           for(var j=0; j < dep_graph[id].length;j++){
                               var dep_id = dep_graph[j]
                               if(inv_dep_graph[dep_id] === undefined){
                                   inv_dep_graph[dep_id] = {}
                               }
                               inv_dep_graph[dep_id].push(id)
                           }
                        }
```

Then in the second loop, you can work on the inv_dep_graph and it should work, no? (Note: maybe my code above is not completely correct, since it is written on github. But I think you can get what I mean.)

**Comment by benelot** [comment]
@On your solution: yes, I think you should not change the second loop too much, since it does what it should. It finds cycles and it decreases the in-degree counters. It is just the wrong dependency graph.

**Comment by aditya95sriram** [comment]
> On your solution: I think you should not change the second loop too much, since it does what it should. You make the processed_nodes counter not count up until you get 0 in-degree, which is not the case anymore in the case of circular deps. The original second loop finds cycles, traverses the graph correctly and it decreases the in-degree counters. It is just the wrong dependency graph.

Yes, I agree. I landed myself in a soup after trying to drastically change something that works well most of the time. 

I'll try to code up the solution from your previous answer (the `inv_dep_graph` one) replacing `cell_ids` with the keys of the `dep_graph` and see if it does the job perfectly.

**Comment by aditya95sriram** [comment]
I tried implementing the aforementioned ideas and I'm confused now. I feel like we're trying to implement a mixture of two solutions, when we should be sticking to one. 

In the following discussion, in/out degree is based on the convention that if `x` depends on `y` (denoted by `x ← y`) `y` is called the parent of `x` (the `in_degree` object we are currently constructing follows this convention). 

There are two perfectly valid ways to perform a topological sort
1. Successively picking vertices with no children and `unshift`ing them to start of the current order
2. Successively picking vertices with no parents and `push`ing them to end of the current order

In Strategy 1, we start with a cell on which no other cell is dependent i.e. we check for `out_degree[id] == 0`, and the starting cell or the root cell conveniently satisfies this criteria. And then decrement the `out_degree` of the nodes on which `id` depends i.e. `dep_graph[id]`. 
In Strategy 2, we start with a cell which doesn't depend on any other cell i.e. we check for `in_degree[id] == 0` (note that the root cell does not satisfy this criteria). And then decrement the `in_degree` of the nodes depending on `id` i.e. `inv_dep_graph[id]`

In the first phase, we compute the `in_degree` which means we have to follow Strategy 2 but then in the second phase, we start by inserting the root cell into the order which implies that we are following Strategy 1. Do you see the point I'm trying to make ? 

A suggested fix would be to pick a strategy (I prefer Strategy 1) and then computing the appropriate `in/out_degree` object and then using the appropriate `(inv)_dep_graph`. Also in Strategy 2, we need to make one pass first to search for the first cell to add to the order (the cell which doesn't depend on any other cell).

**Comment by benelot** [comment]
I see what you mean. The thing is that the dep_graph is basically what we define through the annotations. So, calculating the out-degree is really the easier solution. So, if you can do this, together with checking if there is not cyclic graph, then this is a fine solution. Actually again, the second loop of the code is fine, you just have to adapt how nodes are added to the list describing the order. 

So instead of calculating the in_degrees, we need to calculate the out-degrees. I have the feeling that this is exactly what I tried to do in the beginning, but then there was the problem with the duplicate nodes in the processing queue that messed it up.

**Comment by benelot** [comment]
Actually, my last solution is not the mixture, but just the strategy 2. It is just that we already would have the dep_graph, but I feel out_degrees seem to be harder to calculate.

**Comment by aditya95sriram** [comment]
Okay, what do you suggest I try ?

Should I try implementing `out_degree` (which seems like it can be done in a fashion similar to computing `inv_dep_graph`) and then adapt the second part to work with `out_degree` instead of `in_degree` ?

**Comment by benelot** [comment]
If you want, you can try to implement out_degree, but I feel that it is not so easy to do. But maybe I am wrong. If I have time this evening, I will try to implement the solution I proposed last (I have another machine with a working setup of nbextensions).

**Comment by aditya95sriram** [comment]
Okay, I have managed to implement `out_degree`. I started with the code that is [currently in the repository](https://github.com/ipython-contrib/jupyter_contrib_nbextensions/blob/master/src/jupyter_contrib_nbextensions/nbextensions/execution_dependencies/execution_dependencies.js#L73) and just added the `visited` check, which seems to be computing the `out_degree` correctly (tested on shown example, no deps, single dep and cyclic deps). This makes sense if you go back and look at the comparison table I had in my issue description, it is closer to `out_degree` than to `in_degree`.

But, it's still unable to catch cyclic dependencies (because `processed_nodes > cell_ids.length` is never triggered). This I'm not sure how to solve, while causing minimal change.

Should I create a PR for this current version that computes `out_degree` correctly, and will compute the processing order perfectly in the absence of cycles ? Or should we complete the cycle check thing and then commit ?

**Comment by benelot** [comment]
It is important that the cycle check works. Otherwise it is very
problematic to run it in the browser.

**Comment by aditya95sriram** [comment]
For the cycle check, I see two approaches depending on how the rest of the topological sort algorithm is structured.
1. **Method 1:** There is an outer main loop which adds one node every iteration to the `processing_order` and an inner loop which goes over all the nodes looking for a node with 0 `in_degree`, adding it to the order and decrementing the `in_degree` of its neighbors. In this method, the moment the inner loop fails to find a 0 `in_degree` node, we have detected a cycle. This approach is slightly inefficient, O(n²) to be specific [(slides 6-15)](https://courses.cs.washington.edu/courses/cse326/03wi/lectures/RaoLect20.pdf)
2. **Method 2:** This is method is more efficient because it makes use of queues (and also closely resembles what we have). It keeps adding potential 0 `in_degree` nodes to a queue rather than repeatedly looping over all nodes to find one. It then adds the vertex to the `processing_order` and decrements it's neighbors `in_degree`. Here, the algorithm terminates when the queue becomes empty, which in the existence of a cycle will occur before all nodes have been added to the `processing_order` [(slides 17-20)](https://courses.cs.washington.edu/courses/cse326/03wi/lectures/RaoLect20.pdf). This seems to be the easy to implement, except that it requires the count of the nodes that are dependencies of the current node and not just all nodes i.e. we are only looking at the connected digraph formed by the current cells (not necessarily direct) dependencies. Currently, `dep_graph` contains all the nodes, even the ones which are not dependencies of the currently executing cell. If we fix `dep_graph` to only hold the dependencies of the current cell (or filter it to discard the ones which are not dependencies), then the cycle check can be added with much ease.

**Comment by Alexis-Prel** [comment]
Sorry to cut through the discussion but why not constituting a good old dependency graph?
A very usual approach for these kind of problems is to do the following:

```python
dependencies = {
    "a": [],
    "b": [],
    "e": ["r", ],
    "h": ["r", ],
    "i": ["r", ],
    "start": ["h", "e", "i"],
    "r": ["a", "b"],
}
todo = set(dependencies.keys())
plan = []

# Catch direct circular dependencies
 if any(task in dependencies[task] for task in todo):
    raise ValueError(f"Circular dependencies found.")

# Plan the tasks
while len(plan) < len(todo):
    at_least_one_planned = False
    for task in todo:
        if (task not in plan) and (not dependencies[task]):
            # Plan the task
            plan.append(task)
            at_least_one_planned = True

            # Remove it from everyone else's dependencies
            for parent in dependencies.keys():
                if task in dependencies[parent]:
                    dependencies[parent].remove(task)

     # Catch indirect circular dependencies or missing requirements
     if not at_least_one_planned:
         raise ValueError("Inconsistent or incomplete dependency tree")
print(plan)
```

Output:
`['a', 'b', 'r', 'e', 'i', 'h', 'start']`

**Comment by benelot** [comment]
That is precisely what has to be done. Mind making a pull request out of
this? Then we can close the issue and accept the change to fix the bug.

**Comment by aditya95sriram** [comment]
@Alexis-Prel Yeah, even I thought that what you had written before the edit made sense.

**Comment by Alexis-Prel** [comment]
Oh, great! Can the pull request wait until Friday? This is a very busy week for me.

**Comment by benelot** [comment]
No worries. I just thought that since you brought up the solution, you
would probably be interested to make the PR.
